### PR TITLE
feat: add universal svg loader

### LINKED ISSUE #96: proposal: add option to load custom icons from external sources
We are trying (@antfu and me) to make a universal toolkit to share some logic to allow load svg icons from any source on `unocss` and `unplugin-icons`. 

Instead creating a shared repo to use it on both projects, I think the better place will be here: this way you can also use it for testing before adding new collections or new icons to existing collections.

The logic to be included here, right now, is only supported on `unplugin-icons`:
- load from external sources (network or filesystem): currently only built-in support for filesystem with optional  transform callback for all icons
- load from inline sources: all collections and their icons on the configuration

The main problem here is that all collection/icon  loading logic is not asynchronous: 
- https://github.com/iconify/iconify/blob/master/packages/utils/src/icon-set/get-icon.ts
- https://github.com/iconify/iconify/blob/master/packages/utils/src/icon-set/get-icons.ts

On `unplugin-icons` we load the collections/icons using async tasks:
- https://github.com/antfu/unplugin-icons/blob/main/src/core/modern.ts#L57
- https://github.com/antfu/unplugin-icons/blob/main/src/core/modern.ts#L25

/cc @cyberalien

### PR DESCRIPTION
**Pull request by userquin**
@cyberalien I added `packages/core/src/modern/index.ts` and a few modules to `packages/utils/src/loader` and added some dependencies to both packages, I need to build the repo and add some tests.

/cc @antfu

For both packages also added `"@types/debug": "^4.1.7"` as `dev dependency`.

`packages/core/package.json`:
```json
	"dependencies": {
		"@antfu/utils": "^0.3.0",
		"@iconify/api-redundancy": "^1.0.2",
		"@iconify/types": "^1.0.10",
		"@iconify/utils": "^1.0.16",
		"cross-fetch": "^3.1.4",
		"debug": "^4.3.3",
                "local-pkg": "^0.4.0"
	},
```

and `packages/utils/package.json`:
```json
	"dependencies": {
                "@antfu/utils": "^0.3.0",
		"@iconify/types": "^1.0.12",
                "debug": "^4.3.3"
        },
```

closes #96

### CODE REVIEW

**Comment by userquin** [comment]
I also need to export the new modules on `packages/utils/src/index.ts`, just added a comment on that module.

**Comment by userquin** [comment]
@cyberalien so,  to build the packages to start adding some tests, what's the order?

**Comment by cyberalien** [comment]
I was going to write that it is as simple as running `npm run build`, but then remembered that new version uses custom build script that is also part of package, so it is actually a bit more complex.

1. Build build script (need to be ran only once):

```sh
cd packages/library-builder
npm run build
```

This will build `library-builder` package, which is similar to `tsup`, but also automatically updates exports in package.json and fixes import paths for ES modules, so they could be used in any environment.

2. Build Utils:

```sh
cd packages/utils
npm run build
```

Test: `npm run test`

**Comment by userquin** [comment]
ok, I need to change some system call since I use windows and `tsc` doesn't exist, it is `tsc.cmd`, I have built `@iconify/library-builder` and `@iconify/utils`.

The steps I follow are:
1) `npm run build` from `packages/library-builder`
2) `npm install` from `packages/utils`
3) `npm run link` from root
4) `npm run build` from `packages/utils`

So, for `packages/core` it should be  the same process but once `@iconify/utils` built.

**Comment by userquin** [comment]
@cyberalien can you take a look to the resulting build? Maybe we can omit exporting `camelize` and cia... they are only used on the loader internally.

**Comment by userquin** [comment]
Pending tasks:
- [ ] use cache: on `packages/core/modern/index.ts` module use `cache` stuff instead `_collections` entry
- [ ] add kolorist: add some colors to the new modules
- [ ] add tests 1: add test for `packages/core/modern`
- [ ] add tests 2: add test for new `packages/utils` stuff
- [ ] add tests 3: add usage of `packages/core/modern/index.ts` on some/all `demo` project

**Comment by userquin** [comment]
How can we test the `modern` module on `core` (it will require some `@iconify-json/**` collection and/or full `@iconify/json`)?

Maybe we can also move the `modern` module to `utils`.

/cc @antfu I add to all loaders a `transform/customize` callback, and so we can configure also other attrs and not only the `scale` for svg `width/height`.

File: packages/core/src/modern/index.ts [hunk 1]
```diff
@@ -53,17 +54,15 @@ export async function loadCollection(name: string, autoInstall = false): Promise
...
 	for (const id of ids) {
 		iconData = getIconData(iconSet, id, true);
 		if (iconData) {
 			debug(`${collection}:${id}`);
-			const { attributes, body } = iconToSVG(iconData, {
-				...DefaultIconCustomizations,
-				height: `${scale}em`,
-				width: `${scale}em`,
-			});
+			const customizations: FullIconCustomisations = { ...DefaultIconCustomizations }
```
Comment (userquin) [inline packages/core/src/modern/index.ts:63]: change this, the customize should return the customizations

**Comment by cyberalien** [comment]
Is it ready for testing and merging?

**Comment by userquin** [comment]
@cyberalien happy new new

I need to include/review latest changes we made on `unplugin-icons` to allow more customizations and include them here, I need to review a few things (very busy with unoccss and vitest-dev). The actual impl is ok for you?

**Comment by cyberalien** [comment]
Happy new year to you too!

As for implementation, there are some questions.

Why is some code in `utils` package, some in `core` package? Utils package is for reusable code, which can be used by many packages. Core package is specific to building icon components in this repository. As far as I see, code in Core package does not rely on anything else from that package, so probably was added there by mistake?

I'm not a fan of using too many third party packages, unless they are really needed. In this case I don't see why `debug` is needed. Looks like its for opinionated logging, not part of actual functionality.

If you need additional debug information, such as which package icon was loaded from in `loadCollection()` function added to Core package, why not return an object with both package name and file path? Same issue with parameters that are passed only for use in debug log, such as `collection` in `getCustomIcon()`.

In Utils package function `getCustomIcon` also uses `debug` package for no apparent reason.

Functions `camelize()` and 2 more similar functions in `loader/utils.ts` can be moved to their own file, for example, `misc/strings.ts`. Then they'll be reusable, separate from other functions in that file.

**Comment by userquin** [comment]
There are 2 functions that cannot be moved to utils, since the types are on core, and this packages also depends on utils, so we cannot move it to utils.

Can you do a refactor? Next week I'll try to finish this.

**Comment by cyberalien** [comment]
> There are 2 functions that cannot be moved to utils, since the types are on core, and this packages also depends on utils, so we cannot move them to utils.

Which ones? I don't see any local imports in code added to core. 

> Can you do a refactor? Next week I'll try to finish this.

Sure

**Comment by userquin** [comment]
upps, you are right, only depends on types not on core, so, can we move `modern/index.ts` to `utils`?

**Comment by cyberalien** [comment]
Sure.

I'll merge this, then will refactor stuff within next few days.

**Comment by userquin** [comment]
@cyberalien we also need to include more utilities... so we work now on dev directly?

**Comment by cyberalien** [comment]
Sure.

I've started some refactoring, pushed some stuff in `dev/more-utils` branch.

After playing with `debug` package, I do see its advantages and it only has 1 dependency, so it is fine. However, in `tryInstallPkg()` you are using `console.log`. Wouldn't it make sense to use `debug` in that file as well for consistency?

**Comment by userquin** [comment]
if you need to do some cleanup/update, you can check the original code here https://github.com/antfu/unplugin-icons/blob/main/src/core/utils.ts, on the same directory you have also `modern.ts`

**Pull request merged**